### PR TITLE
Refactor archive support with libarchive

### PR DESCRIPTION
**Pull request by FooIbar**
Fixes [tachiyomi#537](https://git.mihon.tech/tachiyomi/tachiyomi/issues/537), [tachiyomi#3988](https://git.mihon.tech/tachiyomi/tachiyomi/issues/3988), [tachiyomi#10121](https://git.mihon.tech/tachiyomi/tachiyomi/issues/10121)

I think Firebase is the only non-free component now? #758

Test archive 1: 376 MB, 27 P, Deflate ZIP
Test archive 2: 4.99 GB, 1727 P, Deflate ZIP

### Before
| Archive | getPages (ms) | Avg. page load time (ms) |
| --- | --- | --- |
| 1 | 80 | 310 |
| 2 | 828 | 64 |

### After
| Archive | getPages (ms) | Avg. page load time (ms) |
| --- | --- | --- |
| 1 | 61 | 54 |
| 2 | 170 | 67 |

### CODE REVIEW

**Comment by AntsyLich** [comment]
Only remove the string from the base string file. Weblate will pickup the removal and delete the others.

File: source-local/src/commonMain/kotlin/tachiyomi/source/local/io/Format.kt [hunk 1]
```diff
@@ -2,25 +2,22 @@ package tachiyomi.source.local.io
 
 import com.hippo.unifile.UniFile
 import tachiyomi.core.common.storage.extension
+import tachiyomi.source.local.io.Archive.isSupported
```
Comment (AntsyLich) [inline source-local/src/commonMain/kotlin/tachiyomi/source/local/io/Format.kt:5]:
```suggestion
import tachiyomi.source.local.io.Archive.isSupported as isArchiveSupported
```

File: core/common/src/main/kotlin/tachiyomi/core/common/archive/ArchiveReader.kt [hunk 2]
```diff
@@ -0,0 +1,42 @@
...
+                val entry = archive.getNextEntry() ?: break
+                if (entry.name == entryName) {
+                    return archive
+                }
+            }
+        } catch (e: ArchiveException) {
+            archive.close()
+            throw e
+        }
+        archive.close()
```
Comment (AntsyLich) [inline core/common/src/main/kotlin/tachiyomi/core/common/archive/ArchiveReader.kt:33]: use `finally`?

File: source-local/src/commonMain/kotlin/tachiyomi/source/local/io/Archive.kt [hunk 3]
```diff
@@ -1,13 +1,17 @@
 package tachiyomi.source.local.io
 
 import com.hippo.unifile.UniFile
-import tachiyomi.core.common.storage.extension
 
 object Archive {
 
-    private val SUPPORTED_ARCHIVE_TYPES = listOf("zip", "cbz", "rar", "cbr", "epub")
+    private val SUPPORTED_ARCHIVE_TYPES = listOf("zip", "rar", "7z", "tar", "epub")
```
Comment (AntsyLich) [inline source-local/src/commonMain/kotlin/tachiyomi/source/local/io/Archive.kt:7]: `epub` already has a case in `Format.valueOf` seems to be legacy code.
```suggestion
    private val SUPPORTED_ARCHIVE_TYPES = listOf("zip", "rar", "7z", "tar")
```

File: source-local/src/commonMain/kotlin/tachiyomi/source/local/io/Archive.kt [hunk 4]
```diff
@@ -1,13 +1,17 @@
...
 
     fun isSupported(file: UniFile): Boolean {
-        return file.extension in SUPPORTED_ARCHIVE_TYPES
+        val name = file.name?.lowercase() ?: return false
+        val extension = name.substringAfterLast('.')
+        val nameWithoutExtension = name.substringBeforeLast('.')
+        return extension in SUPPORTED_ARCHIVE_TYPES ||
+            extension.startsWith("cb") || // cbz, cbr, etc.
+            nameWithoutExtension.endsWith(".tar") // tar.gz, tar.bz2, etc.
```
Comment (AntsyLich) [inline source-local/src/commonMain/kotlin/tachiyomi/source/local/io/Archive.kt:15]: I'm more inclined to explicitly set which archive are supported.

**Review by AntsyLich — CHANGES_REQUESTED** [review]

File: core/common/src/main/kotlin/tachiyomi/core/common/archive/ArchiveReader.kt (hunk 2 above)
Comment (FooIbar) [reply core/common/src/main/kotlin/tachiyomi/core/common/archive/ArchiveReader.kt:33]: `archive` should not be closed if it's returned.

File: source-local/src/commonMain/kotlin/tachiyomi/source/local/io/Archive.kt (hunk 3 above)
Comment (FooIbar) [reply source-local/src/commonMain/kotlin/tachiyomi/source/local/io/Archive.kt:7]: It's used here https://github.com/mihonapp/mihon/blob/36e40c099772d2cb53d4ec87b2b00f97fe455c98/source-local/src/androidMain/kotlin/tachiyomi/source/local/LocalSource.kt#L257

Comment (AntsyLich) [reply source-local/src/commonMain/kotlin/tachiyomi/source/local/io/Archive.kt:7]: Ok fair enough. Should probably be a separate check tho.

Comment (FooIbar) [reply source-local/src/commonMain/kotlin/tachiyomi/source/local/io/Archive.kt]: https://github.com/mihonapp/mihon/blob/36e40c099772d2cb53d4ec87b2b00f97fe455c98/source-local/src/androidMain/kotlin/tachiyomi/source/local/LocalSource.kt#L188-L191
It's also used here, but I don't think epub will have ComicInfo.xml?

Comment (AntsyLich) [reply source-local/src/commonMain/kotlin/tachiyomi/source/local/io/Archive.kt]: It shouldn't under normal circumstances.

File: core/common/src/main/kotlin/tachiyomi/core/common/archive/ZipWriter.kt [hunk 5]
```diff
@@ -0,0 +1,74 @@
+package tachiyomi.core.common.archive
```
Comment (AntsyLich) [inline core/common/src/main/kotlin/tachiyomi/core/common/archive/ZipWriter.kt:1]:
```suggestion
package mihon.core.common.archive
```

**Review by AntsyLich — CHANGES_REQUESTED** [review]

File: app/src/main/java/eu/kanade/tachiyomi/ui/reader/loader/ArchivePageLoader.kt [hunk 6]
```diff
@@ -0,0 +1,36 @@
...
+            entries
+                .filter { it.isFile && ImageUtil.isImage(it.name) { reader.getInputStream(it.name)!! } }
+                .sortedWith { f1, f2 -> f1.name.compareToCaseInsensitiveNaturalOrder(f2.name) }
+                .mapIndexed { i, entry ->
+                    ReaderPage(i).apply {
+                        stream = { reader.getInputStream(entry.name)!! }
+                        status = Page.State.READY
+                    }
+                }.toList()
+        }
```
[the remark this answers is not in the document]
Comment (FooIbar) [reply app/src/main/java/eu/kanade/tachiyomi/ui/reader/loader/ArchivePageLoader.kt:26]: Such things should be handled by formatter.

File: core/common/src/main/kotlin/eu/kanade/tachiyomi/util/storage/EpubFile.kt [hunk 7]
```diff
@@ -59,9 +38,9 @@ class EpubFile(channel: SeekableByteChannel) : Closeable {
      * Returns the path to the package document.
      */
     fun getPackageHref(): String {
-        val meta = zip.getEntry(resolveZipPath("META-INF", "container.xml"))
+        val meta = getInputStream(resolveZipPath("META-INF", "container.xml"))
         if (meta != null) {
-            val metaDoc = zip.getInputStream(meta).use { Jsoup.parse(it, null, "") }
+            val metaDoc = meta.use { Jsoup.parse(it, null, "") }
```
[the remark this answers is not in the document]
Comment (FooIbar) [reply core/common/src/main/kotlin/eu/kanade/tachiyomi/util/storage/EpubFile.kt:43]: I don't think the suggested change is more readable, and it actually compares to null twice.

File: app/src/main/java/eu/kanade/tachiyomi/ui/reader/loader/ArchivePageLoader.kt (hunk 6 above)
[the remark this answers is not in the document]
Comment (AntsyLich) [reply app/src/main/java/eu/kanade/tachiyomi/ui/reader/loader/ArchivePageLoader.kt:26]: yeah i'll add one later

**Pull request merged**